src/Microsoft.OData.Client/BulkUpdateSaveResult.cs

```diff
@@ -0,0 +1,236 @@
+
+                try
+                {
+                    this.batchResponseMessage = this.RequestInfo.GetSynchronousResponse(bulkUpdateRequestMessage, false);
+                }
+                catch (DataServiceTransportException ex)
+                {
+                    InvalidOperationException exception = WebUtil.GetHttpWebResponse(ex, ref this.batchResponseMessage);
+
+                    throw exception;
+                }
+            }
+        }
+
+        /// <summary>
+        /// This method processes all the changed descriptors in the entity tracker.
+        /// It loops through all the descriptors and creates relationships between the descriptors if any.
+        /// </summary>
+        /// <typeparam name="T">The type of the top-level resources we want to do a deep update on.</typeparam>
+        /// <param name="descriptors">The list of descriptors in the entity tracker.</param>
+        /// <param name="IfRootObject">True if the object is a root object, i.e, it was passed by the client, otherwise false.</param>
+        /// <param name="objects">Objects of the top-level type we want to perform a deep update on.</param>
+        /// <exception cref="InvalidOperationException"></exception>
+        internal void BuildDescriptorGraph<T>(IEnumerable<Descriptor> descriptors, bool ifRootObject, params T[] objects)
+        {
+            foreach (object obj in objects)
+            {
+                EntityDescriptor rootEntityDescriptor = this.RequestInfo.EntityTracker.GetEntityDescriptor(obj);
+                
+                // If we do not find an EntityDescriptor in the entitytracker
+                // for any of the provided objects then we throw an exception. 
+                if (rootEntityDescriptor == null)
+                {
+                    throw Error.InvalidOperation(Strings.Context_EntityNotContained);
+                }
+
+                // If it is a top-level object,
+                // meaning that it was sent as top-level object by the user
+                // then we add it to the top-level objects. 
+                if (ifRootObject)
+                { 
+                    bulkUpdateGraph.AddTopLevelDescriptors(rootEntityDescriptor);
+                }
+
+                foreach (Descriptor descriptor in descriptors)
+                {
+                    if (descriptor.DescriptorKind == DescriptorKind.Entity)
+                    {
+                        EntityDescriptor entityDescriptor = descriptor as EntityDescriptor;
+
+                        if (entityDescriptor.ParentEntity != null)
+                        {
+                            if (entityDescriptor.ParentEntity.Equals(rootEntityDescriptor.Entity))
+                            {
+                                // this check avoids cyclic operations. 
+                                if (!bulkUpdateGraph.CheckIfLinkAlreadyExists(entityDescriptor))

```

Comment:
Since these conditions are relatively straightforward, I think we can merge the if-statements, reduce the indentation and make the code slightly easier to read:

```suggestion
                            if (entityDescriptor.ParentEntity.Equals(rootEntityDescriptor.Entity)
                                 && !bulkUpdateGraph.Contains(entityDescriptor)
                             )
                            {

```